cpp/src/io/json/column_tree_construction.cu

```diff
@@ -0,0 +1,581 @@
+                        non_leaf_nodes.begin(),
+                        non_leaf_nodes_children.begin(),
+                        thrust::equal_to<TreeDepthT>());
+
+  thrust::scatter(rmm::exec_policy_nosync(stream),
+                  non_leaf_nodes_children.begin(),
+                  non_leaf_nodes_children.end(),
+                  non_leaf_nodes.begin(),
+                  rowidx.begin() + 1);
+
+  print<NodeIndexT>(rowidx, "h_rowidx", stream);
+
+  thrust::transform_inclusive_scan(
+    rmm::exec_policy_nosync(stream),
+    thrust::make_zip_iterator(thrust::make_counting_iterator(1), rowidx.begin() + 1),
+    thrust::make_zip_iterator(thrust::make_counting_iterator(1) + num_columns, rowidx.end()),
+    rowidx.begin() + 1,
+    cuda::proclaim_return_type<NodeIndexT>([] __device__(auto a) {
+      auto n   = thrust::get<0>(a);
+      auto idx = thrust::get<1>(a);
+      return n == 1 ? idx : idx + 1;
+    }),
+    thrust::plus<NodeIndexT>{});
+
+  print<NodeIndexT>(rowidx, "h_rowidx", stream);
+
+  rmm::device_uvector<NodeIndexT> colidx((num_columns - 1) * 2, stream);
+  // Skip the parent of root node
+  thrust::scatter(rmm::exec_policy_nosync(stream),
+                  parent_col_ids.begin() + 1,
+                  parent_col_ids.end(),
+                  rowidx.begin() + 1,
+                  colidx.begin());
+  // excluding root node, construct scatter map
+  rmm::device_uvector<NodeIndexT> map(num_columns - 1, stream);
+  thrust::inclusive_scan_by_key(rmm::exec_policy_nosync(stream),
+                                parent_col_ids.begin() + 1,
+                                parent_col_ids.end(),
+                                thrust::make_constant_iterator(1),
+                                map.begin());
+  thrust::for_each_n(rmm::exec_policy_nosync(stream),
+                     thrust::make_counting_iterator(1),
+                     num_columns - 1,
+                     [rowidx         = rowidx.begin(),
+                      map            = map.begin(),
+                      parent_col_ids = parent_col_ids.begin()] __device__(auto i) {
+                       auto parent_col_id = parent_col_ids[i];
+                       if (parent_col_id == 0)
+                         map[i - 1]--;
+                       else
+                         map[i - 1] += rowidx[parent_col_id];
+                     });
+  thrust::scatter(rmm::exec_policy_nosync(stream),
+                  thrust::make_counting_iterator(1),
+                  thrust::make_counting_iterator(1) + num_columns - 1,
+                  map.begin(),
+                  colidx.begin());
+
+  print<size_type>(max_row_offsets, "h_max_row_offsets", stream);
+
+  // Mixed types in List children go to different columns,
+  // so all immediate children of list column should have same max_row_offsets.
+  //   create list's children max_row_offsets array
+  //   gather the max_row_offsets from children row offset array.
+  {
+    auto max_row_offsets_it =
```

Comment:
Have adopted this for `mapped_col_ids_copy`.  Since we are reusing the results of cudf::io::json::detail::reduce_to_column_tree, the lifespan of intermediate results get limited. The rowidx and colidx constructions have also be separated out into smaller lambdas.